### PR TITLE
fix potential memcpy overlap issue in join

### LINKED ISSUE #8796: Potential bug in join when build has many duplicated keys
## Bug Report

Please answer these questions before submitting your issue. Thanks!
In join probe, if build has many duplicated key, the intermedia result block could be large, in order to control the overall memory usage, on input block will be probed multiple times, each time only data in [probe_process_info.start_row, probe_process_info.end_row) will be processed. 
But some data structures used in join assume whole block data is processed, for example:
https://github.com/pingcap/tiflash/blob/ef90ce478aff27a7499a1a55debac5957f326476/dbms/src/Interpreters/Join.cpp#L870
`anti_filter` and `offsets_to_replicate` in `handleOtherConditions` both assumes it contains all the data in block. In order to adopt this, before `handleOtherConditions`, the caller need to shrink `anti_filter` and `offsets_to_replicate`. Currently, it use [`assign`](https://github.com/pingcap/tiflash/blob/ef90ce478aff27a7499a1a55debac5957f326476/dbms/src/Common/PODArray.h#L623). However, `shrink` actually should not use `memcpy` since there is a chance that dst and src have overlap, and it is undefined behavior for `memcpy` is dst and src have overlap

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: close #8796

Problem Summary:

Ref #8796 for details.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Common/PODArray.h [hunk 1]
```diff
@@ -631,6 +631,21 @@ class PODArray : public PODArrayBase<sizeof(T), INITIAL_SIZE, TAllocator, pad_ri
         this->c_end = this->c_start + bytes_to_copy;
     }
 
+    void assignFromSelf(size_t start, size_t end)
+    {
+        if unlikely (start == 0)
+        {
+            return;
+        }
```
Comment (gengliqi) [inline dbms/src/Common/PODArray.h:639]:
```suggestion
        if unlikely (start == 0)
            return;

```

**Review by gengliqi — APPROVED** [review]

File: dbms/src/Common/PODArray.h [hunk 2]
```diff
@@ -631,6 +631,19 @@ class PODArray : public PODArrayBase<sizeof(T), INITIAL_SIZE, TAllocator, pad_ri
...
+
+        assert(end >= start && end <= this->size());
+        size_t required_capacity = end - start;
+        size_t bytes_to_copy = this->byte_size(required_capacity);
+        const auto & from_begin = this->begin() + start;
+        memmove(this->c_start, reinterpret_cast<const void *>(&*from_begin), bytes_to_copy);
+        this->c_end = this->c_start + bytes_to_copy;
+    }
+
     void assign(const PODArray & from) { assign(from.begin(), from.end()); }
```
Comment (yibin87) [inline dbms/src/Common/PODArray.h:647]: Or add self check in assign?

**Review by yibin87 — APPROVED** [review]
LGTM

**Pull request merged**